### ISSUE TITLE
setup preflight:false doesn't work in Preact wmr dev server

### ISSUE THREAD

**Issue by danielweck** [tw-in-js/twind#135]
(Twind version `0.15.8`)

I am able to merge new values into Tailwind's default preflight reset (see example below), but `preflight: false` doesn't work as intended (just in case, I also tried `preflight: 0`, `preflight: null` and `preflight: undefined`, to no avail).

```
setup({
		preflight: (preflight) => {
			console.log(preflight);
			return css({
				h1: {
					color: 'red',
				},
			});
		},
});
```

**Comment by danielweck** [comment]
Documentation: https://github.com/tw-in-js/twind/blob/main/docs/advanced/setup.md#preflight

**Comment by sastan** [comment]
Using `preflight: false` does work. [See this example](aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCmltcG9ydCB7IGRvbVNoZWV0IH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvc2hlZXRzJwoKc2V0dXAoewogIHByZWZsaWdodDogZmFsc2UsCiAgLy8gSnVzdCBmb3IgdGhpcyBkZW1vCiAgc2hlZXQ6IGRvbVNoZWV0KCkKfSkKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxtYWluIGNsYXNzPSJ0dyAke3R3YGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgcC01YH0iPgogICAgPGgxIGNsYXNzPSIke3R3YGZvbnQtYm9sZCB0ZXh0KGNlbnRlciA1eGwgd2hpdGUgc206Z3JheS04MDAgbWQ6cGluay03MDApYH0iPlRoaXMgaXMgVHdpbmQhPC9oMT4KICAgIDxwPkNvbnRlbnQgb2Ygc3R5bGUgdGFnPC9wPgogICAgPHByZT4ke0FycmF5LmZyb20oc2VsZi5fX3R3aW5kLmNoaWxkTm9kZXMsIG4gPT4gbi50ZXh0Q29udGVudCkuam9pbignXG4nKX08L3ByZT4KICAgIDwvbWFpbj4KYA)

**Comment by danielweck** [comment]
Your "See this example" link is broken :)

**Comment by sastan** [comment]
Sorry: [Here you go](https://esm.codes/#aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCmltcG9ydCB7IGRvbVNoZWV0IH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvc2hlZXRzJwoKc2V0dXAoewogIHByZWZsaWdodDogZmFsc2UsCiAgLy8gSnVzdCBmb3IgdGhpcyBkZW1vCiAgc2hlZXQ6IGRvbVNoZWV0KCkKfSkKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxtYWluIGNsYXNzPSJ0dyAke3R3YGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgcC01YH0iPgogICAgPGgxIGNsYXNzPSIke3R3YGZvbnQtYm9sZCB0ZXh0KGNlbnRlciA1eGwgd2hpdGUgc206Z3JheS04MDAgbWQ6cGluay03MDApYH0iPlRoaXMgaXMgVHdpbmQhPC9oMT4KICAgIDxwPkNvbnRlbnQgb2Ygc3R5bGUgdGFnPC9wPgogICAgPHByZT4ke0FycmF5LmZyb20oc2VsZi5fX3R3aW5kLmNoaWxkTm9kZXMsIG4gPT4gbi50ZXh0Q29udGVudCkuam9pbignXG4nKX08L3ByZT4KICAgIDwvbWFpbj4KYA==)

**Comment by danielweck** [comment]
Indeed! I can see it works with the "shim" method too:
https://esm.codes/#aW1wb3J0IHsgc2V0dXAgfSBmcm9tICdodHRwczovL2Nkbi5za3lwYWNrLmRldi90d2luZCcKaW1wb3J0IHsgZG9tU2hlZXQgfSBmcm9tICdodHRwczovL2Nkbi5za3lwYWNrLmRldi90d2luZC9zaGVldHMnCmltcG9ydCB7IGNyZWF0ZU9ic2VydmVyIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvb2JzZXJ2ZSc7CnNldHVwKHsKICBwcmVmbGlnaHQ6IGZhbHNlLAogIC8vIEp1c3QgZm9yIHRoaXMgZGVtbwogIHNoZWV0OiBkb21TaGVldCgpCn0pCmNvbnN0IG9ic2VydmVyID0gY3JlYXRlT2JzZXJ2ZXIoKTsKb2JzZXJ2ZXIub2JzZXJ2ZShkb2N1bWVudC5kb2N1bWVudEVsZW1lbnQpOwoKZG9jdW1lbnQuYm9keS5pbm5lckhUTUwgPSBgCiAgPG1haW4gY2xhc3M9Imgtc2NyZWVuIGJnLXB1cnBsZS00MDAgcC01Ij4KICAgIDxoMSBjbGFzcz0iZm9udC1ib2xkIHRleHQoY2VudGVyIDV4bCB3aGl0ZSBzbTpncmF5LTgwMCBtZDpwaW5rLTcwMCkiPlRoaXMgaXMgVHdpbmQhPC9oMT4KICAgIDxwPkNvbnRlbnQgb2Ygc3R5bGUgdGFnPC9wPgogICAgPHByZT4ke0FycmF5LmZyb20oc2VsZi5fX3R3aW5kLmNoaWxkTm9kZXMsIG4gPT4gbi50ZXh0Q29udGVudCkuam9pbignXG4nKX08L3ByZT4KICAgIDwvbWFpbj4KYA==

I wonder why it doesn't work from PreactWMR?

**Comment by sastan** [comment]
Could you create a sandbox or repo for me to investigate?

**Comment by danielweck** [comment]
Ah! I've narrowed it down to `npm start` with PreactWMR (`preflight: false` works fine with `npm run build && npm run serve`, with or without `--prerender`)

**Comment by danielweck** [comment]
For the avoidance of doubt: I am not using the server-side Twind "shim" ( even with PreactWMR's `--prerender` option), Twind's "shim" only runs on the client side. I tried placing the Twind calls before and after Preact's `hydrate` (which is in fact aliased to `render` when using `npm start`), this makes no difference.

**Comment by danielweck** [comment]
Yep, `config.preflight` is changed from `false` to `undefined` when using `npm start` with PreactWMR:
https://github.com/tw-in-js/twind/blob/ecad9cfb2f12c5bf7905535ab1b1abc59085a302/src/twind/configure.ts#L282-L297
(to find this out, I `console.log()` the value by directly patching the `twind.js` minified code)

**Comment by danielweck** [comment]
PROD (`npm run build && npm run serve`) vs. DEV (`npm start`):

<img width="577" alt="PreactWMR_PROD" src="https://user-images.githubusercontent.com/625201/110048692-42fe2000-7d48-11eb-93bc-97d4487d74d1.png">

<img width="577" alt="PreactWMR_DEV" src="https://user-images.githubusercontent.com/625201/110048710-4b565b00-7d48-11eb-8066-0b75a7ac4c23.png">

**Comment by danielweck** [comment]
As shown in the screenshot above, there are two consecutive calls to the "Twind config" logging, and in the debugger I can see two code bundles (twind and twind/observe). So it looks like PreactWMR is generating incorrect bundles, probably something to do with Hot Module Reload ...

**Comment by danielweck** [comment]
I solved this issue with a workaround, using:
```
import { setup } from 'twind/shim';
	setup({
		hash: false,
		preflight: false,
		modez: 'strict',
	});
```

... instead of:
```
import { setup } from 'twind';
import { createObserver } from 'twind/observe';
const observer = createObserver();
	setup({
		hash: false,
		preflight: false,
		modez: 'strict',
	});
observer.observe(document.body);
```
(which achieves the same but without the "impure" / "side effect" of `twind/shim` which causes `setup()` to be called twice:

https://github.com/tw-in-js/twind/blob/ecad9cfb2f12c5bf7905535ab1b1abc59085a302/src/shim/index.ts#L28

**Comment by danielweck** [comment]
Alright, I found a "proper" fix:

https://unpkg.com/browse/twind@0.15.8/shim/shim.js

=> replace `import{setup as d}from"../twind.js";import{createObserver as c}from"../observe/observe.js";`
with `import{setup as d}from"twind";import{createObserver as c}from"twind/observe";`

https://unpkg.com/browse/twind@0.15.8/observe/observe.js

=> replace `import{tw as h}from"../twind.js";`
with `import{tw as h}from"twind";`

...otherwise the PreactWMR dev bundler creates duplicates of `twind.js`.

Here is a prime example of this pattern with Preact's own Hook library:

https://unpkg.com/browse/preact@10.5.12/hooks/dist/hooks.mjs

=> note `import{options as n}from"preact";` instead of using a relative path to:

https://unpkg.com/browse/preact@10.5.12/dist/preact.mjs

**Comment by danielweck** [comment]
To conclude, I think this is a bug in Twind's published packages. Relative paths shouldn't be used in imports as they are across packages.

The reason why the aforementioned https://esm.codes links work is because they serve Twind via the Skypack CDN. I would be curious to see what other dev-time "bundlers" do (e.g. Vite)

**Comment by sastan** [comment]
Relative paths are constantly used to reference other files within the same package. Honestly, I believe this to be a bug in WMR.

@danielweck Could you provide a codesandbox or repo to show case the issue?

@JoviDeCroock @cristianbote Shouldn't WMR resolve all imports to a unique id (like the absolute) path and it should "detect" that `import "twind"`(-> `node_modules/twind/twind.js`) is the same as `import "../twind.js"` from `node_modules/twind/observe/observe.js`?

**Comment by sastan** [comment]
> I solved this issue with a workaround, using:
> 
> ```
> import { setup } from 'twind/shim';
> 	setup({
> 		hash: false,
> 		preflight: false,
> 		modez: 'strict',
> 	});
> ```
> 
> ... instead of:
> 
> ```
> import { setup } from 'twind';
> import { createObserver } from 'twind/observe';
> const observer = createObserver();
> 	setup({
> 		hash: false,
> 		preflight: false,
> 		modez: 'strict',
> 	});
> observer.observe(document.body);
> ```
> 

@danielweck  Why did you use `createObserver` instead of the shim?

**Comment by danielweck** [comment]
> Relative paths are constantly used to reference other files within the same package.

that's the point, they have dedicated package.json in their own subfolders (see my Preact/hooks example for a counter point to your assertion)

**Comment by danielweck** [comment]
...for example, this is the observe "sub"-package:
https://unpkg.com/browse/twind@0.15.8/observe/package.json

**Comment by danielweck** [comment]
> @danielweck Why did you use `createObserver` instead of the shim?

as mentioned in a previous comment, the shim has a side effect (i.e. execution of code as soon as the module is requires/imported, so unpure module) ... which makes sense in a web browser context when included as a script tag executed early during page loading, but in my Preact app lifecycle the duplicate call to setup and observe.disconnect/connect bothered me (purely on principle), but practically-speaking I was also able unable to defer the twind shim execution for testing (i.e. in relation to SSR and hydrate), so I decided to implement the shim directly in order to control it's lifecycle.

note that my workaround is now different than what I stated earlier: i now have a postinstall script that fixes the erroneous twind/subpackage esm imports (in my opinion ;) instead of reverting back to using the twind shim (I really wanted to avoid the side effects / unpure module)

**Comment by danielweck** [comment]
> @JoviDeCroock @cristianbote Shouldn't WMR resolve all imports to a unique id (like the absolute) path and it should "detect" that ``

reminder (as this thread is getting long): Preact WMR 'npm run build && npm run serve' (with or without --prerender) works fine.

 It is just 'npm run start' that fails because duplicates of twind are created on the fly and loaded into the browser execution context (clearly visible in the debugger).

I am sure the PreactWMR devs will be able to fix their module id logic (if they decide to do so), by establishing the equivalence of twind.js with ../twind.js (for example).

however i believe the correct NPM publish pattern is to not cross the import boundary in sub-packages (like preact + preact-hooks which live in the same top level package but are published with their distinct package.json to advertise their facets)

**Comment by danielweck** [comment]
I wonder if export maps could help here?
https://github.com/preactjs/wmr/blob/b5c9cac534c33b147744655ec707923b1e62212d/packages/wmr/src/plugins/npm-plugin/resolve.js#L88

**Comment by sastan** [comment]
We the same layout and package.json fields (incl exports, main, module, ...) like preact **except** that we are using relative imports from child modules:

- https://unpkg.com/browse/twind@0.15.8/package.json
- https://unpkg.com/browse/twind@0.15.8/shim/package.json
- https://unpkg.com/browse/twind@0.15.8/shim/shim.js

There may be some edge cases where the bare module import (like preact) may have advantages (for example overriding a specific submodule). I still believe that this is an issue within WMR (at least for `npm start`). **But** for compatibility reasons I will change to bare imports from submodules.

**Comment by sastan** [comment]
> ...for example, this is the twind/observe "sub"-package:
> https://unpkg.com/browse/twind@0.15.8/observe/package.json

What is wrong with that? Same as preact. Or did I miss something? I think the difference is the relative import from within sub-package JS file.

**Comment by danielweck** [comment]
> > ...for example, this is the twind/observe "sub"-package:
> > https://unpkg.com/browse/twind@0.15.8/observe/package.json
> 
> What is wrong with that? Same as preact. Or did I miss something? I think the difference is the relative import from within sub-package JS file.

My point exactly (sorry, I was writing from my tiny phone keyboard so I probably wasn't clear enough :)

**Comment by sastan** [comment]
> as mentioned in a previous comment, the shim has a side effect (i.e. execution of code as soon as the module is requires/imported, so unpure module) ... which makes sense in a web browser context when included as a script tag executed early during page loading, but in my Preact app lifecycle the duplicate call to setup and observe.disconnect/connect bothered me (purely on principle), but practically-speaking I was also able unable to defer the twind shim execution for testing (i.e. in relation to SSR and hydrate), so I decided to implement the shim directly in order to control it's lifecycle.

You are right. The twind/shim module is unpure / has side effects. Using twind/shim within test context is still an open issue. If you have any input how we could improve the DX there, please let me know.

**Comment by danielweck** [comment]
> You are right. The twind/shim module is unpure / has side effects.

The problem was hard to diagnose at first because of tree shaking (i.e. an unused import is discarded, but a simple `console.log` on one of the module exports triggers the inclusion of the module in the bundle / execution context).
So I decided to look at Twind's source code and was glad to see that I could re-implement the shim easily thanks to the public observer methods :)

Thanks for your work by the way, much appreciated!

**Comment by sastan** [comment]
Short update: I have a working WMR setup as you have described. Will publish/release soon.

**Comment by sastan** [comment]
Fixed with v0.15.9

**Comment by danielweck** [comment]
Works at my end, thanks for the prompt updates! :)
